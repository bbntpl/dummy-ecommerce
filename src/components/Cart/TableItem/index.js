import TableItem from './TableItem';

export default TableItem;