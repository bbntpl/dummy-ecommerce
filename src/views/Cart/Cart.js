import { useState, useEffect, useCallback } from 'react';
import { Grid, Loader } from 'semantic-ui-react';

import ProductsTable from '../../components/Cart/ProductsTable';
import Checkout from '../../components/Cart/Checkout';
import Placeholder from '../../components/Cart/Placeholder';

import { discountedPrice } from '../../js/reusableFuncs';
import DecimalPrecision from '../../js/decimalPrecision';
import styled from 'styled-components';

const StyledGrid = styled(Grid)`
	positon: relative;
	& .right-col--sticky{
		width: 300px!important;
		& > div {
			margin-top: 30px;
			position: sticky!important;
			top: 0;
			right: 0;
		}
	}
`
export default function Cart(props) {
	const { cart } = props;
	const [isLoading, setIsLoading] = useState(true);
	const [totalCost, setTotalCost] = useState(0);

	const getTotalItems = useCallback(() => {
		return cart.reduce((total, { quantity }) => {
			return total + quantity;
		}, 0);
	}, [cart]);

	useEffect(() => {
		const calculatedTotalCost = (cart) => {
			const rawTotalCost = [...cart].reduce((totalCost, product) => {
				const { price, discountPercentage, quantity } = product;
				const newPrice = discountedPrice(price, discountPercentage);
				return totalCost + (newPrice * quantity);
			}, 0);
			return DecimalPrecision.round(rawTotalCost, 2);
		}
		const newTotalCost = calculatedTotalCost(cart);

		setTotalCost(prevTotalCost =>
			prevTotalCost + (newTotalCost - prevTotalCost));
	}, [cart]);

	// only display loader if cart is empty
	useEffect(() => {
		setIsLoading(false);
	}, [totalCost]);

	return (
		<StyledGrid container stackable>
			{isLoading && <Loader active></Loader>}
			{
				!isLoading && (cart.length && totalCost) ?
					<>
						<Grid.Row columns={2}>
							<Grid.Column width={12} only='computer'>
								<ProductsTable
									{...props}
									totalCost={totalCost}
									getTotalItems={getTotalItems}
								/>
							</Grid.Column>
							<Grid.Column width={10} only='tablet'>
								<ProductsTable
									{...props}
									totalCost={totalCost}
									getTotalItems={getTotalItems}
								/>
							</Grid.Column>
							<Grid.Column only='mobile'>
								<ProductsTable
									{...props}
									totalCost={totalCost}
									getTotalItems={getTotalItems}
								/>
							</Grid.Column>
							<Grid.Column width={4} only='computer' className='right-col--sticky'>
								<Checkout
									totalCost={totalCost}
									getTotalItems={getTotalItems}
								/>
							</Grid.Column>
							<Grid.Column
								width={6}
								only='tablet mobile'
								className='right-col--sticky'
								floated='right'
							>
								<Checkout
									totalCost={totalCost}
									getTotalItems={getTotalItems}
								/>
							</Grid.Column>
						</Grid.Row>
					</>
					: <Placeholder />
			}
		</StyledGrid>
	)
}