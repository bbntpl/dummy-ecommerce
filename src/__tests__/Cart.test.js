import { render, screen } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import Cart from '../views/Cart';

const cartItems = [
	{
		id: 1,
		title: 'title1',
		price: 10.10,
		discountPercentage: 20,
		stock: 8,
		thumbnail: null,
		quantity: 4,
	},
	{
		id: 2,
		title: 'title2',
		price: 10.11,
		discountPercentage: 30,
		stock: 5,
		thumbnail: null,
		quantity: 2,
	},
];

describe('conditionally render components by value of passed prop', () => {
	it('should render empty card placeholder', () => {
		const { container } = render(
			<Router><Cart cart={[]} /></Router>
		);

		expect(container).toMatchSnapshot();
	})

	it('should render cart features', () => {
		render(
			<Router>
				<Cart cart={cartItems} />
			</Router>
		);

		const totalItems = screen.getAllByText(/6 items/i);
		expect(totalItems).toBeTruthy();
	})
})

