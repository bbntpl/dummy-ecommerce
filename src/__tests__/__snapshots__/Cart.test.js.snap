// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`conditionally render components by value of passed prop should render empty card placeholder 1`] = `
<div>
  <div
    class="ui container stackable grid sc-kDDrLX dZdSWl"
  >
    <div
      class="sc-jqUVSM bZUpzn"
    >
      <img
        class="ui image sc-papXJ jUyhQW"
        src="cart-img-placeholder.png"
      />
      <div
        class="ui header"
      >
        Your cart is empty
      </div>
      <a
        href="/shop"
      >
        Go shopping
      </a>
    </div>
  </div>
</div>
`;
